cirq/pasqal/pasqal_device.py

```diff
@@ -0,0 +1,208 @@
+# limitations under the License.
+from typing import FrozenSet, Iterable, cast
+from numpy import sqrt
+
+import cirq
+
+from cirq.pasqal import ThreeDGridQubit
+
+
+@cirq.value.value_equality
+class PasqalDevice(cirq.devices.Device):
+    """A Pasqal Device with qubits placed on a 3D grid."""
+
+    def __init__(self, control_radius: float,
+                 qubits: Iterable[ThreeDGridQubit]) -> None:
+        """ Initializes the description of the device.
+
+        Args:
+            control_radius: the maximum distance between qubits for a controlled
+                gate. Distance is measured in units of the indices passed into
+                the qubit constructor.
+            qubits: Qubits on the device, identified by their x, y, z position.
+                Must be of type ThreeDGridQubit.
+
+        Raises:
+            ValueError: if the wrong qubit type is provided or if invalid
+                parameter is provided for control_radius. """
+        us = cirq.value.Duration(micros=1)
+
+        self._measurement_duration = 5000 * us
+        self._gate_duration = 2 * us
+        self._max_parallel_z = 2
+        self._max_parallel_xy = 2
+        self._max_parallel_c = 10
+        self._max_parallel_t = 1
+
+        for q in qubits:
+            if not isinstance(q, ThreeDGridQubit):
+                raise TypeError('Unsupported qubit type: {!r}'.format(q))
+
+        if not control_radius >= 0:
+            raise ValueError("control_radius needs to be a non-negative float")
+
+        self.control_radius = control_radius
+
+        self.qubits = qubits
+
+    def qubit_set(self) -> FrozenSet[cirq.Qid]:
+        return frozenset(self.qubits)
+
+    def qubit_list(self):
+        return [qubit for qubit in self.qubit_set()]
+
+    def decompose_operation(self,
+                            operation: cirq.ops.Operation) -> 'cirq.OP_TREE':
+
+        decomposition = [operation]
+
+        if not isinstance(operation, cirq.ops.GateOperation):
+            raise TypeError("{!r} is not a gate operation.".format(operation))
+
+        # Try to decompose the operation into elementary device operations
+        if not PasqalDevice.is_pasqal_device_op(operation):
+            decomposition = cirq.protocols.decompose(
+                operation, keep=PasqalDevice.is_pasqal_device_op)
+
+        for dec in decomposition:
+            # coverage: ignore
+            if not PasqalDevice.is_pasqal_device_op(dec):
+                raise TypeError("Don't know how to work with {!r}.".format(
+                    operation.gate))
+
+        return decomposition
+
+    @staticmethod
+    def is_pasqal_device_op(op: cirq.ops.Operation) -> bool:
+        if not isinstance(op,
+                          (cirq.ops.GateOperation, cirq.ParallelGateOperation)):
+            return False  # coverage: ignore
+
+        keep = False
+
+        # Currently accepting all multi-qubit operations
+        keep = keep or (len(op.qubits) > 1)
```

Comment:
It seems like `keep` is always false in this method. Why are we `or`ing it with the gate instances? Is this variable necessary?

Reply:
I think keep actually changes from False to True, when the statement on the right is True.

Reply:
Oh! I see. Hmm. There's gotta be a way we can covey this in less lines. Something like

```python
return  (len(op.qubits) > 1) or (isinstance(op.gate, cirq.ops.YPowGate)) or (isinstance(op.gate, cirq.ops.ZPowGate)) or (isinstance(op.gate, cirq.ops.XPowGate)) or (isinstance(op.gate, cirq.ops.PhasedXPowGate)) or (isinstance(op.gate, cirq.ops.MeasurementGate)) or (isinstance(op.gate, cirq.ops.IdentityGate))
```
?